// omni-ui/tools/external-codegen/plopfile.mjs
import process from 'node:process';

const root = process.cwd();
export default function (plop) {
  /** @type {import('plop').NodePlopAPI} */
  plop.setHelper('getName', name => {
    const pathCheck = name.split('/');
    const checkForHyphen = pathCheck[pathCheck.length - 1].split('-');
    let transformedName = checkForHyphen.map(value => {
      return value.charAt(0).toUpperCase() + value.substring(1);
    });
    transformedName = transformedName.join('');
    const checkExtension = transformedName.split('.');
    if (checkExtension.length > 1) return checkExtension[0];
    return transformedName;
  });
  plop.setHelper('getTagName', name => {
    const hyphenCheck = name.split('-');
    if (hyphenCheck.length > 1) {
      const pathCheck = name.split('/');
      const extensionCheck = pathCheck[pathCheck.length - 1].split('.');
      return extensionCheck[0];
    } else {
      throw 'Tag name must contain hyphen!!';
    }
  });
  plop.setHelper('getPath', name => {
    const extensionCheck = name.split('.');
    return extensionCheck.length === 1 ? `${name}.js` : `${name}`;
  });
  plop.setHelper('getTemplatePath', name => {
    const extensionCheck = name.split('.');
    if (extensionCheck.length === 1) {
      return './component-js.hbs';
    } else {
      if (extensionCheck[extensionCheck.length - 1].toLowerCase() === 'js') return './component-js.hbs';
      else if (extensionCheck[extensionCheck.length - 1].toLowerCase() === 'ts') return './component-ts.hbs';
      else throw 'Invalid extension';
    }
  });
  plop.setGenerator('create-component', {
    description: 'Omni-UI Component',
    prompts: [
      {
        type: 'input',
        name: 'name',
        message: "Enter your component tag name, e.g.,'omni-element'"
      }
    ],
    actions: [
      {
        type: 'add',
        path: `${root}/{{getPath name}}`,
        templateFile: '{{getTemplatePath name}}'
      }
    ]
  });
}
